database/rrdhost.c

```diff
@@ -674,12 +674,20 @@ void rrdhost_free(RRDHOST *host) {
     rrd_check_wrlock();     // make sure the RRDs are write locked
 
     // ------------------------------------------------------------------------
-    // clean up the sender
+    // clean up streaming
     rrdpush_sender_thread_stop(host); // stop a possibly running thread
     cbuffer_free(host->sender->buffer);
     buffer_free(host->sender->build);
     freez(host->sender);
     host->sender = NULL;
+    netdata_mutex_lock(&host->receiver_lock);
+    if (host->receiver) {
+        netdata_thread_cancel(host->receiver->thread);
+        destroy_receiver_state(host->receiver);
```

Comment:
This deallocates the data structures before `host->receiver->thread` is gone, and it corrupts the memory every time, and potentially during runtime, not only during shutdown.